perma_web/api2/serializers.py

```diff
@@ -0,0 +1,242 @@
+        return super(BaseSerializer, self).update(instance, validated_data)
+
+
+### LINKUSER ###
+
+class LinkUserSerializer(BaseSerializer):
+    full_name = serializers.ReadOnlyField(source='get_full_name')
+    short_name = serializers.ReadOnlyField(source='get_short_name')
+    top_level_folders = serializers.SerializerMethodField()
+
+    class Meta:
+        model = LinkUser
+        fields = ('id', 'first_name', 'last_name', 'full_name', 'short_name', 'top_level_folders')
+
+    def get_field_names(self, *args, **kwargs):
+        # Exclude the top_level_folders field when this serializer is nested inside another serializer.
+        # This lets us use the same serializer at /user and also to embed created_by user info in /archives
+        field_names = super(LinkUserSerializer, self).get_field_names(*args, **kwargs)
+        if self.parent:
+            field_names = [i for i in field_names if i != 'top_level_folders']
+        return field_names
+
+    def get_top_level_folders(self, user):
+        serializer = FolderSerializer(user.top_level_folders(), many=True)
+        return serializer.data
+
+
+### FOLDER ###
+
+class FolderSerializer(BaseSerializer):
+    has_children = serializers.SerializerMethodField()
+
+    class Meta:
+        model = Folder
+        fields = ('id', 'name', 'parent', 'has_children', 'organization')
+        extra_kwargs = {'parent': {'required': True}}
+        allowed_update_fields = ['name', 'parent']
+
+    def get_has_children(self, folder):
+        return not folder.is_leaf_node()
+
+    def validate_name(self, name):
+        if self.instance:
+            # renaming
+            if self.instance.is_shared_folder:
+                raise serializers.ValidationError("Shared folders cannot be renamed.")
+            if self.instance.is_root_folder:
+                raise serializers.ValidationError("User's main folder cannot be renamed.")
+        return name
+
+    def validate_parent(self, parent):
+        if self.instance:
+            # moving
+            if self.instance.is_shared_folder:
+                raise serializers.ValidationError("Can't move organization's shared folder.")
+            if self.instance.is_root_folder:
+                raise serializers.ValidationError("Can't move user's main folder.")
+        return parent
+
+    def validate(self, data):
+        if 'name' in data:
+            # make sure folder name is unique in this location
+            parent_id = data['parent'].pk if 'parent' in data else self.instance.parent_id if self.instance else None
+            if parent_id:
+                unique_query = Folder.objects.filter(parent_id=parent_id, name=data['name'])
+                if self.instance:
+                    unique_query = unique_query.exclude(pk=self.instance.pk)
+                if unique_query.exists():
+                    raise serializers.ValidationError({'name':"A folder with that name already exists at that location."})
+        return data
```

Comment:
Similar question as above: since name is a required field (default behavior), I would have thought we could rely on it always being present in data, without an explicit check. Not the case?